### ISSUE TITLE
Clear Text Credentials in kong.conf

### ISSUE THREAD

**Issue by rudijs** [Mashape/kong#1806]
Hi,

Kong API Gateway is a recent addition to our stack, so the organization has conducted a security review from a third party.

They've flagged as 'high risk' that we have clear text credentials (Posgresql) in /usr/loca/kong/kong.conf

This file is auto generated on kong start and is 0644 permissions.

So it's an issue I need to address, else the organization security department could exclude Kong from our stack.

I believe there is no Kong auto reconnect to the database, so after Kong has successfully started, I can have a script update the /usr/local/kong/kong.conf file and replace or remove the pgsql creds.

This does raise the question of Kong restart, in our case we're using Kong in Openshift w/ Kubernetes so in order to restart Kong the entire container is restarted, which recreates kong.conf.

This proposed solution, for our case, means at run time there are no creds in kong.conf, only at startup which is the only time a database connection is made.

I hope I explained that OK, if not please let me know.

Make reasonable sense do you think?

- Kong version 0.9.3
- Operating System OpenShift v1.1.4 with the official Kong docker image

**Comment by thefosk** [comment]
Instead of putting the password in the configuration file, you can set them in an ENV variable, ie:

```
KONG_PG_PASSWORD=secret123
```

And maybe you can do something like this:

```shell
KONG_PG_PASSWORD=secret123 kong start && unset KONG_PG_PASSWORD
```

I mean, I suppose this would be a pretty common problem with many other software, how are you currently dealing with it? 

You could also encrypt the HDD at the OS level if that's an option (ie, [EBS encrypted volumes](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html) if you are using AWS)

**Comment by rudijs** [comment]
@thefosk Hey Marco,

Actually we do start Kong with the KONG_PG_PASSWORD set as an environment var, but the /usr/local/kong/kong.conf file is written to disk during the start up process.

So it's like it transfers the pwd env var from env to file.

When our Kong is running in a docker container, the pwd is actually in clear text in both env and file.

The file pwd is what's been highlighted in our security test.

**Comment by rudijs** [comment]
Some more context for this...

Start up the offical kong docker image, but do not start kong yet:

`docker run --rm -it --name kong kong:0.9.3 bash`

List kong config files

`ls /usr/local/kong`

There are no files yes, as kong has not yet started.

Start kong and run it as a background process, set the pgsql host and password

```
KONG_PG_HOST=192.168.33.100 KONG_PG_PASSWORD=P@ssw0rd kong start &
[1] 7
sh-4.2# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0    192     4 ?        Ss+  05:45   0:00 /usr/local/bin/dumb-init /bin/bash /docker-entrypoint.sh sh
root         5  0.0  0.0  11776  1760 ?        Ss   05:45   0:00 sh
root         7  1.0  0.1  40964  7060 ?        S    05:46   0:00 perl /usr/local/openresty/bin/resty /usr/local/bin/kong start
root         8  1.3  0.1  52324  5864 ?        S    05:46   0:00 /usr/local/openresty/bin/../nginx/sbin/nginx -p /tmp/AeRaIiGpqw/ -c conf/nginx.conf
root        43  0.0  0.0  15552   844 ?        S    05:46   0:00 /usr/sbin/dnsmasq -p 8053 --pid-file=/usr/local/kong/pids/dnsmasq.pid -N -o --listen-address=127.0.0.1
root        50  0.0  0.1  14728  5744 ?        Sl   05:46   0:00 serf agent -profile wan -rpc-addr 127.0.0.1:7373 -event-handler member-join,member-leave,member-failed,member-update
root        73  0.0  0.0  11636  1340 ?        S    05:46   0:00 sh -c /usr/local/openresty/nginx/sbin/nginx -p /usr/local/kong -c nginx.conf >"/tmp/lua_BGL9Wk" 2>"/tmp/lua_EQGW3v"
root        74  0.0  0.0  36408  2632 ?        S    05:46   0:00 /usr/local/openresty/nginx/sbin/nginx -p /usr/local/kong -c nginx.conf
root        75  0.0  0.0  47424  1652 ?        R    05:46   0:00 ps aux
```

List kong config files

```
ls /usr/local/kong
client_body_temp  fastcgi_temp  kong.conf  logs  nginx-kong.conf  nginx.conf  pids  proxy_temp  scgi_temp  serf  ssl  uwsgi_temp
```

List the pgsql specific configs from kong.config

```
grep pg /usr/local/kong/kong.conf 
pg_password = P@ssw0rd
pg_ssl_verify = false
pg_ssl = false
pg_database = kong
pg_host = 192.168.33.100
pg_port = 5432
pg_user = kong
```

So this illustrates the issue raised by our third party security audit.

pg_password is clear text in a config file. It's gone from environment variable to persistent text file.

Our security audit flagged this as 'high', which is not good for my team and I.

I'm thinking that now with Kong up and running, with no pgsql reconnect feature, we can remove the pg_password from kong.conf.

In our case with Docker, we don't 'restart' kong, we recreate the container which would recreate the kong.conf w/ pg_password on restart.

Makes sense I think? Thoughts?

Thanks.

**Comment by thibaultcha** [comment]
@rudijs First of all, thank you for reporting this concern and taking the time to describe it in details. The reason why this file is written in the prefix is because the CLI being a different application than Kong itself, we must use a way to pass the computed configuration (defaults mixed with user config file and ENV variables) to OpenResty. Thus, we write a "compiled" version of the configuration to the Nginx prefix, so that when it starts, we read this file to load our configuration. 

Now, indeed, we could remove the file once that is done since we do not want such secrets to be publicly available, your concern is valid. However, the reason why that isn't done was originally because `kong reload` (which is the equivalent of sending a `HUP` signal to Nginx, will spawn new workers that need this configuration.

One solution that I have been considering (I am not a fan of this whole approach to write a compiled config and store it in the prefix), was to use a lua shared dictionary to store the configuration, since such dictionaries are the only memory zones surviving `HUP` signals. I believe doing so would also resolve this concern at the same time, since the credentials would not be stored in plain text anymore.

**Comment by rudijs** [comment]
@thibaultcha OK cool, that all makes sense.

Thanks also for the detailed response, is does bring absolute clarity as to why it is how it is.

So, in our case, a tradeoff could be that we remove the kong.conf file after kong start up - and then no be able to use SIGHUP.

This can work OK for us as we're using Kubernetes (inside Openshift) so we'll just spin up a new container, direct traffic to it, and terminate the old container.

In effect it's the same 'type' of thing - a kong restart.

I'll implement this approach and see if its all good for us.

I'll report my results here when I have them after the weekend.

I'll close this issue for now - Thanks!

**Comment by Tieske** [comment]
@rudijs so what were your results? if you don't mind sharing? we had another user with a similar question, see https://gitter.im/Mashape/kong?at=58783830300f220a66df91d2
(please respond here for future reference, and not on Gitter)

Thx!

**Comment by rudijs** [comment]
@Tieske Non-conclusive results to far, I need to re-test.

My only test was in non-prod, which seemed fine.

In prod there was some weirdness, which I was really able to debug fully - was under pressure to just roll back and get prod working again.

I'm not sure if there was some condition, maybe an error, that cause kong/nginx to want to re-read the config file (which I had removed).

Moving forward, I'll need to re-test the theory again.

It's still an issue for us, our security audit tagged this as a high issue on their report.

**Comment by Tieske** [comment]
I'll reopen this for now

**Comment by Tieske** [comment]
@rudijs thx for the feedback!

**Comment by hutchic** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [plain text secrets](https://trello.com/c/T6vYHiF4/32-plain-text-secrets)

**Comment by rudijs** [comment]
@Tieske We've been running in production for about a week now with the kong.conf file removed after start up.

We're using the official Kong docker image in Openshift (Kubernetes).

60 seconds after startup we remove the file /usr/local/kong/kong.conf

We don't 'reload' the config, rather we redeploy the container, which creates kong.conf for startup, then we remove after 60s.

Credentials are still in the shell environment which we need to address with a different approach to startup.

**Comment by p0pr0ck5** [comment]
@rudijs just a heads up, we merged in #2611 a change to how Kong write the config file; this is now removes "other" read permissions, addressing part of your concern here.

The reality of completely removing clear-text credentials from the file system is something that we'll never get away from, and it's not the appropriate solution anyhow. A few good resources on this:

https://t-rob.net/2011/10/24/encrypting-passwords-in-config-files-secure-or-not/
https://security.stackexchange.com/questions/15040/standards-for-encrypting-passwords-in-configuration-files/15055#15055

Defining credentials in plaintext in a config file isn't the problem domain here- it's limiting the scope of access for such a configuration.
